enhancements/installer/component-selection.md

```diff
@@ -0,0 +1,423 @@
+the `Default` value (include or exclude).
+
+The `Default` field (carried through to the CVO) allows an Admin to choose what behavior they want for new
+capabilities that might come as part of a future upgrade (either have the new capability installed during
+the upgrade, or exclude it)
+
+Specifying the same capability as an `include` and an `exclude` is an error condition.
+
+
+Examples of candidate components to be treated as capabilities:
+
+* console
+* imageregistry
+* samples
+* cluster baremetal operator
+* olm/marketplace
+* csi-*
+* insights
+* monitoring
+* ???
+
+
+
+3) Pass the list of filtered annotations to the CVO.  This is distinct from overrides because overrides
+put the cluster in an unsupported state.  Filtered annotations are supported for upgrades.  The filtered
+components will be listed in the [ClusterVersion](https://github.com/openshift/api/blob/4436dc8be01e8dcd8b250e1b32bb0fbd64ba78ac/config/v1/types_cluster_version.go#L35) object:
+
+```yaml
+  spec:
+    capabilities:
+      default: [include|exclude]
+      exclude:
+      - console
+      - samples
+      include:
+      - registry
+```
+
+The CVO will filter out(not apply/reconcile) resources that are annotated with `capability.openshift.io/$exclusions`
+and listed in the `exclude` section, or are not listed and the `default` setting is `exclude`.  In addition
+any explicitly included capability should be added, regardless of the CLUSTER_PROFILE.
+
+If a resource has multiple `capability.openshift.io` annotations, then the resource will be filtered if any
+of the annotations on the resource are matched by a configured filter.  This allows a component to define
+a `ServiceMonitor` which is part of both the `Monitoring` capability and the `Foo` capability.  If either
+of those capabilities are excluded, the `ServiceMonitor` should not be created.  (If Monitoring is excluded
+then no ServiceMonitors should be created, if only Foo is excluded, then only Foo's ServiceMonitors should
+be excluded).
+
+Specifying the same capability as an `include` and an `exclude` is an error condition.
+
+Note: The CVO must wait until the ClusterVersion resource exists (created by the installer) to ensure
+there is no race condition in which the CVO starts creating resources before it has the filter list.
+
+4) Admin can remove an item from the excluded annotations list, but they cannot add an item to it.  If an
+item is removed, the CVO will apply the previously filtered(skipped) resources to the cluster on the next reconciliation.
+Adding an item to the filtered list is not supported because it requires the component be removed from the
+running cluster which has more significant implications for how all traces of the component are removed.  The
+CVO should reject admin edits/updates that add new entries to the exclusion list.
+
+The currently configured capabilities settings for the CVO should be recorded in telemeter so we can understand
+the configuration of a given cluster.
```

Comment:
[`status.capabilities`][1] would get this information up via Insights, although with both Insights and Monitoring in the list of optional capabilities, we are likely to have a few clusters unable to report either way.  Do you want both/and reporting, to try and cover clusters that enable Monitoring but disable Insights?

[1]: https://github.com/openshift/enhancements/pull/922#discussion_r741252865

Reply:
we can probably start w/ just doing it via monitoring/telemeter since i doubt monitoring will be the first capability we make optional.  Might be worth revisiting adding insights-based reported if/when we make monitoring an optional capability though.

Reply:
well, as long as you aren't worried about "there's bogus data in ClusterVersion's `spec`" (e.g. because there's an admission webhook you really trust, or you're doing Red-Hat-side Insights sanity checking), you wouldn't actually need new code for the Insights path.  Because Insights tarballs always include the full ClusterVersion YAML already.  If Insights is also not going to be one of the first components to go conditional, perhaps we can defer the new metric/Telemetry until later?